app/grandchallenge/cases/models.py

```diff
@@ -948,3 +980,87 @@ def start_dicom_import_job(self):
             self._mark_failed(
                 error_message="An unexpected error occurred", exc=e
             )
+
+    def get_job_output_manifest(self, *, import_job):
+        output_uri = import_job["outputS3Uri"]
+        parsed = urlparse(output_uri)
+        bucket = parsed.netloc
+        key = parsed.path.lstrip("/") + "job-output-manifest.json"
+
+        # Try to get the manifest.
+        try:
+            obj = self._s3_client.get_object(bucket=bucket, key=key)
+            return obj["Body"]
+        except self._s3_client.exceptions.NoSuchKey as e:
+            raise RetryStep("Manifest not (yet) found for job output") from e
+
+    def get_image_sets_for_dicom_import_job(self, *, import_job):
+        """
+        Retrieves the image sets created for an import job.
+        """
+
+        manifest = self.get_job_output_manifest(import_job=import_job)
+
+        try:
+            data = json.load(manifest)
+        except json.decoder.JSONDecodeError:
+            return []
+
+        expression = jmespath.compile(
+            "jobSummary.imageSetsSummary[].imageSetId"
+        )
+        image_sets = expression.search(data)
+
+        return image_sets
+
+    def handle_event(self, *, event):
+        job_status = event["jobStatus"]
+        try:
+            if job_status == "COMPLETED":
+                self.handle_completed_job(event=event)
+            elif job_status == "FAILED":
+                self.handle_failed_job(event=event)
+                raise ValueError("Import job failed")
+            else:
+                raise ValueError("Invalid job status")
+        except Exception as e:
+            self._mark_failed(
+                error_message="An unexpected error occurred", exc=e
+            )
+            raise
+        else:
+            self.status = self.DICOMImageSetUploadStatusChoices.COMPLETED
+            self.save()
+
+    def handle_completed_job(self, *, event):
+        image_sets = self.get_image_sets_for_dicom_import_job(import_job=event)
+        if not image_sets:
+            raise ValueError("No image sets created.")
+        valid = self.validate_image_sets(image_sets)
+        if not valid:
+            self.cleanup_image_sets(image_sets)
+            raise ValueError("Image sets invalid.")
+        self.convert_image_sets_to_internal(image_sets)
+
+    def handle_failed_job(self, *, event):
+        image_sets = self.get_image_sets_for_dicom_import_job(import_job=event)
+        if image_sets:
+            self.cleanup_image_sets(image_sets)
```

Comment:
For failed jobs, you also want to know the exact error message and hence would need to inspect the failure.ndjson, right? Even if we don't want to expose it to participants, we might want to log it to sentry given that the S3 bucket contents get deleted eventually, right?